src/ripple/overlay/MessageScheduler.h

```diff
@@ -0,0 +1,496 @@
+        NetClock::duration timeout);
+
+    // TODO: Stop using shared_ptr for messages.
+    // Instead, pass a unique_ptr to the receiver.
+    // We don't need it after they're done.
+    // They can either hold it or let it destruct.
+    void
+    receive(std::shared_ptr<protocol::TMLedgerData> message);
+    void
+    receive(std::shared_ptr<protocol::TMGetObjectByHash> message);
+
+    // TODO: Let callers withdraw senders and receivers.
+
+    void
+    stop();
+
+private:
+    /**
+     * Offer channels to senders, in turn, until senders either
+     * (a) close no channels,
+     * in which case we skip over them, or
+     * (b) stop scheduling new senders,
+     * in which case they are effectively removed.
+     *
+     * @pre `channels` is not empty.
+     * @post Either `channels` is empty,
+     * or every sender in `senders` refused to close any channel.
+     */
+    void
+    _offer(std::vector<Channel>& channels, std::vector<Sender*>& senders);
+
+    template <typename Message>
+    void
+    _send(
+        std::shared_ptr<Peer> peer,
+        RequestId requestId,
+        Message const& message,
+        Receiver* receiver,
+        NetClock::duration timeout)
+    {
+        return _send(
+            peer,
+            requestId,
+            message,
+            protocolMessageType(message),
+            receiver,
+            timeout);
+    }
+
+    void
+    _send(
+        std::shared_ptr<Peer> peer,
+        RequestId requestId,
+        ::google::protobuf::Message const& message,
+        protocol::MessageType type,
+        Receiver* receiver,
+        NetClock::duration timeout);
+
+    void
+    _receive(
+        RequestId requestId,
+        std::shared_ptr<::google::protobuf::Message> message);
+};
+
+constexpr const char*
+toString(MessageScheduler::FailureCode code)
+{
+    switch (code)
+    {
+        case MessageScheduler::FailureCode::TIMEOUT:
+            return "TIMEOUT";
+        case MessageScheduler::FailureCode::DISCONNECT:
+            return "DISCONNECT";
+        case MessageScheduler::FailureCode::SHUTDOWN:
+            return "SHUTDOWN";
+        default:
+            return "<unknown failure code>";
+    }
+}
+
+inline std::ostream&
+operator<<(std::ostream& out, MessageScheduler::FailureCode code)
+{
+    return out << toString(code);
+}
+
+/**
+ * `Offer` represents an offer to close M among N channels, M <= N.
+ * M is called the "size".
+ *
+ * `Offer` is an interface around a set of channels represented by
+ * a `std::vector` of weak pointers.
+ * `Offer` does not own the set; it holds the set by reference.
+ * The set is owned by the stack.
+ * When passed an `Offer`, a `Sender` must use it or lose it.
+ * `Sender`s may not save references to `Channel`s, or make copies.
+ *
+ * `Sender`s may close channels in the offer by sending messages to them.
+ * After `Offer` is destroyed, the set is left with only the remaining
+ * open channels.
+ */
+class MessageScheduler::Offer
```

Comment:
I'd like to avoid names that overlap with business logic terms. Things like "offer" or "withdrawal" probably need different names.